pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1214,4 +1243,86 @@ private int getMaxNumPartitionsPerInstance(InstancePartitions instancePartitions
       return (numPartitions + numInstancesPerReplicaGroup - 1) / numInstancesPerReplicaGroup;
     }
   }
+
+  /**
+   * Validate the committed low level consumer segments to see if its segment store copy is available. Fix the missing segment store copy by asking servers to upload to segment store.
+   * Since uploading to segment store involves expensive compression step (first tar up the segment and then upload), we don't want to retry the uploading. Segment without segment store copy can still be downloaded from peer servers.
+   * @see <a href="https://cwiki.apache.org/confluence/display/PINOT/By-passing+deep-store+requirement+for+Realtime+segment+completion#BypassingdeepstorerequirementforRealtimesegmentcompletion-Failurecasesandhandling">By-passing deep-store requirement for Realtime segment completion:Failure cases and handling</a>
+   */
+  public void uploadToSegmentStoreIfMissing(TableConfig tableConfig) {
+    Preconditions.checkState(!_isStopping, "Segment manager is stopping");
+
+    String realtimeTableName = tableConfig.getTableName();
+    // Get all the LLC segment ZK metadata for this table
+    List<LLCRealtimeSegmentZKMetadata> segmentZKMetadataList = ZKMetadataProvider.getLLCRealtimeSegmentZKMetadataListForTable(_propertyStore, realtimeTableName);
+
+    // Iterate through llc segments and upload missing segment store copy by following steps:
+    //  1. Ask servers which have online segment replica to upload to segment store. Servers return segment store download url after successful uploading.
+    //  2. Update the llc segment ZK metadata by adding segment store download url.
+    for (LLCRealtimeSegmentZKMetadata segmentZKMetadata : segmentZKMetadataList) {
```

Comment:
We should think of a different way to do this. Getting all the segment metadata for all tables every time this task runs can overload zookeeper.

Reply:
one way to reduce the zk read cost is to use `getLLCRealtimeSegments` as suggested in the java doc.
LLC segment name includes segment creation time, we can fix only recent segments (say the last 1 or 2 days)

Reply:
I thought of that, but what if the fix was unsuccessful for multple attempts? (we can always give up, of course).

Another way is to keep a separate queue in zk of all the segments that are in peer state. This is more tricky than it sounds (e.g. we may complete the segment with "peer" url and fail to add to this queue), so I am not certain.

Another way to do this is to maintain an in-memory queue in the controller. Update the queue each time we have a new segment that is in this state.  When the controller starts up, fetch all segments(or recent segments) and populate the queue.

Reply:
The in-memory queue plus a scan of recent segments should fix most segment download url issues. Pinot controllers in our experiences restarted rarely.

Reply:
Added a in memory-queue for to cache the segment names which require fix.